transports/quic/src/endpoint.rs

```diff
@@ -100,75 +90,71 @@ pub struct Endpoint {
     /// See [`Endpoint::new_connections`] (just below) for a commentary about the mutex.
     to_endpoint: Mutex<mpsc::Sender<ToEndpoint>>,
 
-    /// Channel where new connections are being sent.
-    /// This is protected by a futures-friendly `Mutex`, meaning that receiving a connection is
-    /// done in two steps: locking this mutex, and grabbing the next element on the `Receiver`.
-    /// The only consequence of this `Mutex` is that multiple simultaneous calls to
-    /// [`Endpoint::poll_incoming`] are serialized.
-    new_connections: Mutex<mpsc::Receiver<Connection>>,
-
     /// Copy of [`Endpoint::to_endpoint`], except not behind a `Mutex`. Used if we want to be
     /// guaranteed a slot in the messages buffer.
     to_endpoint2: mpsc::Sender<ToEndpoint>,
 
-    /// Socketaddr of the local UDP socket passed in the configuration at initialization after it
-    /// has potentially been modified to handle port number `0`.
-    pub(crate) local_addr: SocketAddr,
+    socket_addr: SocketAddr,
 }
 
 impl Endpoint {
-    /// Builds a new `Endpoint`.
-    pub fn new(config: Config) -> Result<Arc<Endpoint>, io::Error> {
-        let local_socket_addr = match crate::transport::multiaddr_to_socketaddr(&config.multiaddr) {
-            Some(a) => a,
-            None => panic!(), // TODO: Err(TransportError::MultiaddrNotSupported(multiaddr)),
-        };
+    /// Builds a new `Endpoint` that is listening on the [`SocketAddr`].
+    pub fn new_bidirectional(
+        config: Config,
+        socket_addr: SocketAddr,
+    ) -> Result<(Arc<Endpoint>, mpsc::Receiver<Connection>), transport::Error> {
+        let (new_connections_tx, new_connections_rx) = mpsc::channel(1);
+        let endpoint = Self::new(config, socket_addr, Some(new_connections_tx))?;
+        Ok((endpoint, new_connections_rx))
+    }
 
-        // NOT blocking, as per man:bind(2), as we pass an IP address.
-        let socket = std::net::UdpSocket::bind(&local_socket_addr)?;
-        // TODO:
-        /*let port_is_zero = local_socket_addr.port() == 0;
-        let local_socket_addr = socket.local_addr()?;
-        if port_is_zero {
-            assert_ne!(local_socket_addr.port(), 0);
-            assert_eq!(multiaddr.pop(), Some(Protocol::Quic));
-            assert_eq!(multiaddr.pop(), Some(Protocol::Udp(0)));
-            multiaddr.push(Protocol::Udp(local_socket_addr.port()));
-            multiaddr.push(Protocol::Quic);
-        }*/
+    /// Builds a new `Endpoint` that only supports outbound connections.
+    pub fn new_dialer(config: Config) -> Result<Arc<Endpoint>, transport::Error> {
+        let socket_addr = SocketAddrV4::new(Ipv4Addr::LOCALHOST, 0);
```

Comment:
Just double checking here, does the use of `IPv4Addr::LOCALHOST` allow to dial endpoints not colocated on the same host?